### ISSUE TITLE
Document deployment

### ISSUE THREAD

**Issue by pepijndevos** [ninjastorms/ninjastorms#5]
The readme mentions `make deploy`, but nowhere in your makefile is such a target defined. Your wiki mentions a [Boot Process](https://github.com/ninjastorms/ninjastorms/wiki/Boot-Process) page, which does not exist. Your readme also refers to a `docs` folder with more documentation and a`tools` folder for building the toolchain which are nowhere to be found.

In short, all the information about actually generating the u-boot script and binary seem to be missing.

**Comment by pepijndevos** [comment]
Oh, the information is there alright, but was for some reason deleted when moving to automake. Why was it deleted, and what prompted the move to automake?

**Comment by niccokunzmann** [comment]
@pepijndevos Could you tell me, where you found the information and where you would expect it?

**Comment by pepijndevos** [comment]
I just browsed the repo at d5b28ee5722e87b0b7ff61d6bfdd99819d6fc3a9 and there is a Makefile with a deploy target and a `build_toolchain.sh` script. Documentation about the boot process would still be helpful, but the deploy target was the main missing link.

**Comment by niccokunzmann** [comment]
If you like to use it and you get it running again, you could also create a PR, so we can merge it as it runs for you.

**Comment by pepijndevos** [comment]
Any idea about the most recent working commit? Probably I'll first try my own "hello world" from scratch for educational purposes, and see from there.

**Comment by oaken-source** [comment]
Yes, the documentation regarding the build process is out of date, due to some horribly misguided attempts of mine to generalize the build and to support additional architectures. I really should have used a dedicated branch for this...
The latest commit in master should still work, though.

Anyway - the toolchain build script has been moved to a separate repository: https://github.com/ninjastorms/toolchain-generator

and the build should boil down to a simple `./autogen.sh && ./configure && make`

anything beyond that should hopefully become clear by reading the configure.ac and Makefile.in files.

and @niccokunzmann is right, pull requests to fix documentation are always welcome.

**Comment by oaken-source** [comment]
And I think make deploy was not in there anymore because it is not clear what deploy means in the context. what are you deploying? and where? it did not make much sense in a more general context. If I recall correctly, I used to test the build in a qemu-system-arm environment, but I forgot how the actual call looked like. for deployment info to the actual hardware, reading the history of the Makefile seems like the sensible thing to do - there used to be targets to create the u-boot scripts and compile them and things like that.

I really need to tidy things up here, I apologize for the mess.

**Comment by pepijndevos** [comment]
Do you intend to support other targets besides qemu and EV3? It seems doing so would greatly increase the complexity and scope of the project.

Some initial observations while browsing the code/documentation:

In your Makefile you disassemble the binary to get the entry point. This whole step could be handled with a proper ld script, as seen on [OSdev / Raspbery Pi Bare Bones / linking the kernel](http://wiki.osdev.org/Raspberry_Pi_Bare_Bones#Linking_the_Kernel)

In your wiki I see notes about missing references when certain compilation options are used. This could probably be resolved by [Linking libgcc](http://wiki.osdev.org/Libgcc). I believe this is the raison d'être of the toolchain build script?

Do you have any thoughts about user-space and code loading? It seems currently the demos are directly compiled into the kernel. I'm still wrapping my head around how this all works, beyond the mundane part of writing a FAT driver and copying the code to RAM. It seems some syscall mechanism is required.

**Comment by oaken-source** [comment]
I was toying with the idea of an x86 build, but that turned out to be out of scope for this particular project...

I agree that we need a linker script. Could you create a separate issue for that?

the toolchain build script started as an idea to introduce new team members to cross compilers. I don't remember much of the build problems we had, but I think we didn't want to link against libgcc for ... reasons. maybe this could also be discussed further in another issue.

separating some sort of user and kernel mode is planned for future iterations. this system originated from - and is occasionally used in - lectures and seminars. the last iteration of this seminar added a very simple scheduler and a task concept.

**Comment by pepijndevos** [comment]
Taking a closer look at the deploy target, the relevant part to me is not the mounting/copying, but rather the generation of `boot.scr`, which seems like it would be perfectly fine to implement with automake. I need to do a bit of learning about automake though.

Although it seems to me that with the linker file in place, it could just be generated, (documented) and checked in. The entry point seems to be the only dynamic part about it.

**Comment by oaken-source** [comment]
Is everything done here, or should we keep this open?

**Comment by pepijndevos** [comment]
The information is there, but could still be improved. I put the U-boot stuff in the previously non-existing "Boot Process" page, but the rest of the wiki and the readme have not been updated.

The wiki is kind of in a mixed state, but if the new deployment using the uImage is added to the readme, we can close this, as far as I'm concerned.
